UI/MapWnd.cpp

```diff
@@ -5922,7 +5923,11 @@ void MapWnd::PushWndStack(std::shared_ptr<GG::Wnd> wnd) {
 }
 
 void MapWnd::RemoveFromWndStack(std::shared_ptr<GG::Wnd> wnd) {
-    auto it = std::find(m_wnd_stack.begin(), m_wnd_stack.end(), wnd);
+    auto it = std::find_if(m_wnd_stack.begin(), m_wnd_stack.end(),
+        [&wnd](std::weak_ptr<GG::Wnd> weak_wnd) {
+            return weak_wnd.lock() == wnd;
```

Comment:
what happens if an empty pointer is passed in as `wnd` ?

Reply:
The first `weak_ptr` that compares equal to an empty `shared_ptr` will be removed; i.e., any null ones or ones whose `shared_ptr` has already been destructed. If they are null (or have had their `shared_ptr` destructed), they don't contribute anything to the stack anymore, so no harm done.

Reply:
@wheals you could try `GG::ProcessThenRemoveExpiredPtrs()`, which will process a container of weak pointers with a supplied function, while removing any null ones.  There are examples of its use in "FleetWnd.cpp", including [here](https://github.com/freeorion/freeorion/blob/master/UI/FleetWnd.cpp#L438)

Reply:
I think the problem with that function for this use case is that it doesn't give fine-grained enough control over the vector. In specific, both times I iterate over `m_wnd_stack` (`RemoveFromWndStack` and `ReturnToMap`), I remove items from the vector, and doing so during the iteration would invalidate all the iterators.

Reply:
I see.  I did not look at your code closely enough.

You could process the vector afterwards to remove the null `weak_ptrs`.  Otherwise, they just leak.

Reply:
Good idea; they wouldn't have leaked generally because they are removed in ReturnToMap if one presses escape enough, but if someone happens to not use the escape shortcut it could be a problem.